flink-autoscaler/src/main/java/org/apache/flink/autoscaler/RestApiMetricsCollector.java

```diff
@@ -78,19 +91,87 @@ protected Map<FlinkMetric, AggregatedMetric> queryAggregatedVertexMetrics(
                                     EmptyRequestBody.getInstance())
                             .get();
 
-            return responseBody.getMetrics().stream()
-                    .collect(
-                            Collectors.toMap(
-                                    m -> metrics.get(m.getId()),
-                                    m -> m,
-                                    (m1, m2) ->
-                                            new AggregatedMetric(
-                                                    m1.getId() + " merged with " + m2.getId(),
-                                                    Math.min(m1.getMin(), m2.getMin()),
-                                                    Math.max(m1.getMax(), m2.getMax()),
-                                                    // Average can't be computed
-                                                    Double.NaN,
-                                                    m1.getSum() + m2.getSum())));
+            return aggregateByFlinkMetric(metrics, responseBody);
         }
     }
+
+    protected Map<FlinkMetric, AggregatedMetric> queryTmMetrics(Context ctx) throws Exception {
+        try (var restClient = ctx.getRestClusterClient()) {
+            var metricNames = getTmMetricNames(restClient, ctx);
+            var metricNameMapping = new HashMap<String, FlinkMetric>();
+
+            REQUIRED_TM_METRICS.forEach(
+                    fm -> {
+                        var name =
+                                fm.findAny(metricNames)
+                                        .orElseThrow(
+                                                () ->
+                                                        new RuntimeException(
+                                                                "Could not find required TM metric "
+                                                                        + fm.name()));
+                        metricNameMapping.put(name, fm);
+                    });
```

Comment:
I'm a little bit concerned these non-parallel requests might take too long for deployments with hundreds of task managers. Especially because this feature cannot be turned off.

Reply:
In the same way as we query the vertex metrics, we only query aggregated task manager metrics through the job manager. So it's a single query regardless of the number of TMs

Reply:
Oh I missed that. Thanks for clarifying.